filtering.py

```diff
@@ -0,0 +1,597 @@
+                        gain[..., i, :],
+                        x[t, ..., i, None, :],
+                        y[t, ..., j]
+                    )
+
+    return y, v, R
+
+
+def wiener(
+    targets_spectrograms: torch.Tensor,
+    mix_stft: torch.Tensor,
+    iterations: int = 1,
+    use_softmask: bool = True,
+    residual: Optional[str] = None,
+    scale_factor: float = 10.0,
+    eps: float = 1e-07
+):
+    """Wiener-based separation for multichannel audio.
+
+    The method uses the (possibly multichannel) spectrograms `v` of the
+    sources to separate the (complex) Short Term Fourier Transform `x` of the
+    mix. Separation is done in a sequential way by:
+
+    * Getting an initial estimate. This can be done in two ways: either by
+      directly using the spectrograms with the mixture phase, or
+      by using :func:`softmask`.
+
+    * Refinining these initial estimates through a call to
+      :func:`expectation_maximization`.
+
+    This implementation also allows to specify the epsilon value used for
+    regularization. It is based on [1]_, [2]_, [3]_, [4]_.
+
+    References
+    ----------
+    .. [1] S. Uhlich and M. Porcu and F. Giron and M. Enenkl and T. Kemp and
+        N. Takahashi and Y. Mitsufuji, "Improving music source separation based
+        on deep neural networks through data augmentation and network
+        blending." 2017 IEEE International Conference on Acoustics, Speech
+        and Signal Processing (ICASSP). IEEE, 2017.
+
+    .. [2] A. Nugraha and A. Liutkus and E. Vincent. "Multichannel audio source
+        separation with deep neural networks." IEEE/ACM Transactions on Audio,
+        Speech, and Language Processing 24.9 (2016): 1652-1664.
+
+    .. [3] A. Nugraha and A. Liutkus and E. Vincent. "Multichannel music
+        separation with deep neural networks." 2016 24th European Signal
+        Processing Conference (EUSIPCO). IEEE, 2016.
+
+    .. [4] A. Liutkus and R. Badeau and G. Richard "Kernel additive models for
+        source separation." IEEE Transactions on Signal Processing
+        62.16 (2014): 4298-4310.
+
+    Parameters
+    ----------
+
+    targets_spectrograms: torch.Tensor
+                        [shape=(nb_frames, nb_bins, {1,nb_channels}, nb_sources)]
+        spectrograms of the sources. This is a nonnegative tensor that is
+        usually the output of the actual separation method of the user. The
+        spectrograms may be mono, but they need to be 4-dimensional in all
+        cases.
+
+    mix_stft: torch.Tensor [complex, shape=(nb_frames, nb_bins, nb_channels, 2)]
+        STFT of the mixture signal.
+
+    iterations: int [scalar]
+        number of iterations for the EM algorithm
+
+    use_softmask: boolean
+        * if `False`, then the mixture phase will directly be used with the
+          spectrogram as initial estimates.
+
+        * if `True`, a softmasking strategy will be used as described in
+          :func:`softmask`.
+
+    residual: str or None
```

Comment:
@aliutkus why is this `str` or `None`?

Reply:
because it's the name you want for rhe residual source, eg accompaniment, or None if we don't do residual

Reply:
yes, but I don't see any use for the key str in the `wiener` function

https://github.com/sigsep/open-unmix-pytorch/blob/f5e30d1602830eae227fbe6cc03069587b4ddf9e/filtering.py#L493-L499

I guess this was meant to be used when the output would be dict. So I would propose to just make it a `bool`.

Reply:
@aliutkus I changed this in 08a18ab. Please object if you think this is not good